### PR TITLE
Dockerfile-hermes: Add Dockerfile for eid-hermes project

### PR DESCRIPTION
**Pull request by sbates130272**
Add a Dockerfile that pulls the Eideticom QEMU repo and builds the x86
system emulator. It uses a rootfs for this VM that is generated using
gen-images (using cloud-init) and spins the whole thing up.

Note that in order for this Dockerfile to work you need to put some
SSH keys into the docker/hermes folder and ensure the Ubunutu focal
cloud image is also in that folder. This requirement may go away once
we open this project to the public.

Note that the cloud-init networking should allow for instance SSH into
the VM via port 2222. The username and password for the main account
on the VM is 'hermes' and 'zeus'. Note we add a .dockerignore to avoid
pulling the large images into context. We also include a
docker-compose.yml for easier management and a README.md to summarize
this approach.

### CODE REVIEW

File: docker/hermes/gen-images [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  - name: hermes
+    lock_passwd: false
+    sudo: ALL=(ALL) NOPASSWD:ALL
+    groups: users, admin
+    home: /home/hermes
+    shell: /bin/bash
+    plain_text_passwd: 'zeus'
+packages:
+  - build-essential
+  - emacs-nox
```
Comment (iomartin) [inline docker/hermes/gen-images:31]: Can you add ansible as well?

File: docker/hermes/gen-images [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  ens3:
+     dhcp4: true
+     # default QEMU userspace network
+     gateway4: 10.0.2.2
+     nameservers:
+       addresses: [ 10.0.2.3,8.8.8.8 ]
+EOF
+
+if [ ! -f focal-server-cloudimg-amd64.img ]; then
+   wget https://cloud-images.ubuntu.com/focal/current/focal-server-cloudimg-amd64.img
```
Comment (iomartin) [inline docker/hermes/gen-images]: We should put this on the images directory

File: docker/hermes/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+WORKDIR ${HERMES}/ubpf
+RUN git init && \
+  git remote add stephen https://github.com/sbates130272/ubpf.git && \
+  git fetch stephen && \
+  git checkout -b hermes stephen/stdbool-fix
+RUN make -C vm && \
+  make -C vm install
+
+  # Clone QEMU at the point we want and then configure and build
+  # it. Pass in this SHA as an argument.
```
Comment (iomartin) [inline docker/hermes/Dockerfile]: nit: whitespace at the end of line

File: docker/hermes/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+WORKDIR _build
+RUN ../configure --target-list=x86_64-softmmu --prefix=/usr/local/ \
+  --enable-trace-backends=simple --enable-libubpf && \
+  make -j $(nproc) && \
+  make install
+
+  # Now set up the image files folder needed to run the hermes
+  # VM. Note that we will do a volume mount on the host so that these
+  # images persist across docker container spin-ups.
+
```
Comment (iomartin) [inline docker/hermes/Dockerfile]: nit: whitespace on empty line

File: docker/hermes/README.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+3. ```docker-compose up --build``` - optionally builds and then spins
+up a docker container that runs the eid-hermes based VM. It also opens
+port 2222 for ssh into the VM from outside the container. Note you can
+add a ``-d`` to this command to run the container as a background
+task. **Note that you will want to change the HERMES_QEMU_SHA argument
+to point to the commit SHA you care about. Do this in
+docker-compose.yml**.
+
+4. ```ssh -v -p 2222 hermes@localhost``` - note you might have to redo
```
Comment (iomartin) [inline docker/hermes/README.md]: We don't need the -v

File: docker/hermes/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+  # Clone QEMU at the point we want and then configure and build
+  # it. Pass in this SHA as an argument. 
+
+ARG HERMES_QEMU_SHA
+RUN mkdir -p ${HERMES}/qemu
+WORKDIR ${HERMES}/qemu
+RUN git init && \
+  git remote add eid git@github.com:Eideticom/qemu.git && \
+  git fetch eid && \
```
Comment (iomartin) [inline docker/hermes/Dockerfile]: We could do

    git fetch eid $HERMES_QEMU_SHA --depth=1

to limit the amount of data that we need to pull

File: docker/hermes/.gitignore [hunk 7]
```diff
@@ -0,0 +1 @@
+focal-server-cloudimg-amd64.img
```
Comment (iomartin) [inline docker/hermes/.gitignore]: We should add `id_rsa` here as well

File: docker/hermes/docker-compose.yml [hunk 8]
```diff
@@ -0,0 +1,17 @@
+version: "3"
+services:
+
+  hermes-vm:
+    build:
+      dockerfile: ./Dockerfile
+      context: ./
+      args:
+        HERMES_QEMU_SHA: 421feca1573f3b25b0c89d11752c7a7d0cd91f81
```
Comment (iomartin) [inline docker/hermes/docker-compose.yml]: Let's remember to update this before we merge

**Review by iomartin — CHANGES_REQUESTED** [review]
Looks mostly good, just some minor comments

File: docker/hermes/Dockerfile (hunk 6 above)
Comment (sbates130272) [reply docker/hermes/Dockerfile]: @iomartin I changed this a bit. I moved the _SHA to _REF and now the default is the HEAD of the hermes branch. This allows a user to put a branch or a commit SHA in the docker-compose file. I also updated the README to inform users they can alter this in docker-compose to test other branches and commits.

Comment (sbates130272) [reply docker/hermes/Dockerfile]: Oh except the head of hermes branch is broken until we get the libubpf code in. So I will change it to the head of your branch for now.

Comment (iomartin) [reply docker/hermes/Dockerfile]: With _REF we should still be able to use --depth=1. We just needs to drop the `-b` on the checkout:

```
git fetch eid $HERMES_QEMU_REF --depth=1 && \
git checkout $HERMES_QEMU_REF
```

Also, if REF is a branch, then it should not have the `eid/` prefix for this to work

File: docker/hermes/README.md [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+have to do this step once. The rootfs will presist across container
+builds unless you rerun this script. If you do rerun this script you
+will lose any local changes you have made to the rootfs.
+
+3. ```docker-compose up --build``` - optionally builds and then spins
+up a docker container that runs the eid-hermes based VM. It also opens
+port 2222 for ssh into the VM from outside the container. Note you can
+add a ``-d`` to this command to run the container as a background
+task. **Note that you will want to change the HERMES_QEMU_REF argument
+to point to the commit REF you care about. Do this in
```
Comment (iomartin) [inline docker/hermes/README.md]: s/commit REF/commit REF or branch/

File: docker/hermes/gen-images [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+     nameservers:
+       addresses: [ 10.0.2.3,8.8.8.8 ]
+EOF
+
+if [ ! -f ${DIR}/focal-server-cloudimg-amd64.img ]; then
+    wget https://cloud-images.ubuntu.com/focal/current/focal-server-cloudimg-amd64.img \
+	 -O ${DIR}/focal-server-cloudimg-amd64.img
+fi
+
+mkdir -p ${DIR}
```
Comment (iomartin) [inline docker/hermes/gen-images]: The mkdir has to be before the previous if

File: docker/hermes/docker-compose.yml [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+services:
+
+#b5aab992a5a6faf500b91beb2225dec85ace5314
+
+  hermes-vm:
+    build:
+      dockerfile: ./Dockerfile
+      context: ./
+      args:
+        HERMES_QEMU_REF: eid/martin/hermes_v3
```
Comment (iomartin) [inline docker/hermes/docker-compose.yml]: see comment about git fetch

**Review by iomartin — CHANGES_REQUESTED** [review]

File: docker/hermes/Dockerfile (hunk 6 above)
Comment (sbates130272) [reply docker/hermes/Dockerfile]: @iomartin I thought this failed for me. But it looks like it works so I will do this.

File: docker/hermes/docker-compose.yml (hunk 8 above)
Comment (sbates130272) [reply docker/hermes/docker-compose.yml]: Now set to martin/hermes_v3. We can update to hermes once your PR is merged.

**Review by iomartin — APPROVED** [review]

**Pull request merged**